### Задание 1

Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

Дополнительная часть (необязательная)
Выполнить задания 1-3 для всего набора данных.

In [1]:
# импортируем нужные библиотеки
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# сразу при чтении файла заменяем все знаки "?" на пропущенные значения (NaN) и даем имена столбцам

data = pd.read_csv('horse_data.csv', na_values='?',  names = ['surgery', 'age', 'case_num', 'rec_temp', 
                                              'pulse', 'respiratory_rate', 'temp_of_extremities', 
                                              'peripheral_pulse', 'mucous_membranes', 
                                              'capillary_refill_time', 'pain', 'peristalsis', 
                                              'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux',
                                              'nasogastric_reflux_PH', 'rectal_examination_feces', 'abdomen', 
                                              'packed_cell_volume', 'total_protein', 'abdominocentesis_appearance', 
                                              'abdomcentesis_total_protein', 'outcome', 'surgical_lesion', 
                                              '25_type_of_lesion','26_type_of_lesion', '27_type_of_lesion', 'cp_data' ])
#data.info()
data.tail()

surgery  age  case_num  rec_temp  pulse  respiratory_rate  \
295      1.0    1    533886       NaN  120.0              70.0   
296      2.0    1    527702      37.2   72.0              24.0   
297      1.0    1    529386      37.5   72.0              30.0   
298      1.0    1    530612      36.5  100.0              24.0   
299      1.0    1    534618      37.2   40.0              20.0   

     temp_of_extremities  peripheral_pulse  mucous_membranes  \
295                  4.0               NaN               4.0   
296                  3.0               2.0               4.0   
297                  4.0               3.0               4.0   
298                  3.0               3.0               3.0   
299                  NaN               NaN               NaN   

     capillary_refill_time  ...  packed_cell_volume  total_protein  \
295                    2.0  ...                55.0           65.0   
296                    2.0  ...                44.0            NaN   
297                    1.0  ...                60.0            6.8   
298                    1.0  ...                50.0            6.0   
299                    NaN  ...                36.0           62.0   

     abdominocentesis_appearance  abdomcentesis_total_protein  outcome  \
295                          NaN                          NaN      3.0   
296                          3.0                          3.3      3.0   
297                          NaN                          NaN      2.0   
298                          3.0                          3.4      1.0   
299                          1.0                          1.0      3.0   

     surgical_lesion  25_type_of_lesion  26_type_of_lesion  27_type_of_lesion  \
295                2               3205                  0                  0   
296                1               2208                  0                  0   
297                1               3205                  0                  0   
298                1               2208                  0                  0   
299                2               6112                  0                  0   

     cp_data  
295        2  
296        1  
297        2  
298        1  
299        2  

[5 rows x 28 columns]

In [3]:
# убираем ненужные (невыбранные столбцы)
data = data.drop(['respiratory_rate', 'peripheral_pulse','mucous_membranes',
                  'capillary_refill_time', 'pain', 'peristalsis','abdominal_distension', 
                  'nasogastric_tube', 'nasogastric_reflux','nasogastric_reflux_PH', 
                  'rectal_examination_feces', 'abdomen', 'packed_cell_volume','abdominocentesis_appearance',
                  'abdomcentesis_total_protein','surgical_lesion', '25_type_of_lesion',
                  '26_type_of_lesion', '27_type_of_lesion', 'cp_data' ], axis=1)
data.head()

surgery  age  case_num  rec_temp  pulse  temp_of_extremities  \
0      2.0    1    530101      38.5   66.0                  3.0   
1      1.0    1    534817      39.2   88.0                  NaN   
2      2.0    1    530334      38.3   40.0                  1.0   
3      1.0    9   5290409      39.1  164.0                  4.0   
4      2.0    1    530255      37.3  104.0                  NaN   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [4]:
# Выбранные столбцы - 'surgery', 'age', 'case_num', 'rec_temp', 'pulse', 'temp_of_extremities','total_protein', 'outcome'
data.head()

surgery  age  case_num  rec_temp  pulse  temp_of_extremities  \
0      2.0    1    530101      38.5   66.0                  3.0   
1      1.0    1    534817      39.2   88.0                  NaN   
2      2.0    1    530334      38.3   40.0                  1.0   
3      1.0    9   5290409      39.1  164.0                  4.0   
4      2.0    1    530255      37.3  104.0                  NaN   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [5]:
# ИЩЕМ КОЛИЧЕСТВО ПРОПУЩЕННЫХ ЗНАЧЕНИЙ В ДАННОМ ДАТАСЕТЕ 
# использую value.counts чтобы найти количество пропусков и др. значений
print('surgery', data['surgery'].value_counts(dropna = False).reset_index())
# есть один пропуск
print()
print('age', data['age'].value_counts(dropna = False).reset_index())
#у нас 2 варианта значения : "1" соответсвует значению Adult (взрослая лошадь), второе значение должно быть "2". но
# второе встречающееся значение - "9" (не "2"), которое по логике предполагается для молодой лошади (Young)
# поскольку третьего варианта не дано, будет считать, что под "9" скрывается значение "Young"
print()
print('case_num', data['case_num'].value_counts(dropna = False).reset_index())
print('case_num', data['case_num'].nunique())
# Поскольку некоторые лошади обрабатываются более 1 раза, номера могут совпадать (из 300 - 284 уникальных номера)
print()
print('rec_temp', data['rec_temp'].value_counts(dropna = False).reset_index())
# имеется 60 пропущенных значений (из 300)
print()
print('pulse', data['pulse'].value_counts(dropna = False).reset_index())
# имеется 24 пропущенных значений (из 300)
print()
print('temp_of_extremities', data['temp_of_extremities'].value_counts(dropna = False).reset_index())
# имеется 56 пропущенных значений (из 300)
print()
print('total_protein', data['total_protein'].value_counts(dropna = False).reset_index())
# имеется 33 пропущенных значений (из 300)
print()
print('outcome', data['outcome'].value_counts(dropna = False).reset_index())
# имеется только одно пропущенное значение 

surgery    index  surgery
0    1.0      180
1    2.0      119
2    NaN        1

age    index  age
0      1  276
1      9   24

case_num        index  case_num
0     529796         2
1     529424         2
2    5279822         2
3     527916         2
4     530526         2
..       ...       ...
279   530255         1
280   530254         1
281   533836         1
282   530251         1
283   535043         1

[284 rows x 2 columns]
case_num 284

rec_temp     index  rec_temp
0     NaN        60
1    38.0        26
2    38.5        19
3    38.3        18
4    37.8        17
5    38.2        16
6    38.1        12
7    38.6        12
8    37.5        12
9    38.4        11
10   37.6         7
11   37.7         7
12   37.2         7
13   37.9         7
14   38.7         7
15   37.3         6
16   38.8         6
17   39.2         5
18   38.9         4
19   39.5         4
20   39.0         4
21   39.3         4
22   37.4         3
23   37.1         3
24   39.4         3
25   36.5         2


In [6]:
# ИЩЕМ частоту пропущенных количественных значений 
print('rec_temp', data['rec_temp'].value_counts(dropna = False, normalize = True).reset_index())
print()
print('pulse', data['pulse'].value_counts(dropna = False, normalize = True).reset_index())
print()
print('total_protein', data['total_protein'].value_counts(dropna = False, normalize = True).reset_index())

rec_temp     index  rec_temp
0     NaN  0.200000
1    38.0  0.086667
2    38.5  0.063333
3    38.3  0.060000
4    37.8  0.056667
5    38.2  0.053333
6    38.1  0.040000
7    38.6  0.040000
8    37.5  0.040000
9    38.4  0.036667
10   37.6  0.023333
11   37.7  0.023333
12   37.2  0.023333
13   37.9  0.023333
14   38.7  0.023333
15   37.3  0.020000
16   38.8  0.020000
17   39.2  0.016667
18   38.9  0.013333
19   39.5  0.013333
20   39.0  0.013333
21   39.3  0.013333
22   37.4  0.010000
23   37.1  0.010000
24   39.4  0.010000
25   36.5  0.006667
26   39.1  0.006667
27   37.0  0.006667
28   40.3  0.006667
29   36.9  0.003333
30   36.4  0.003333
31   40.8  0.003333
32   36.8  0.003333
33   36.0  0.003333
34   39.6  0.003333
35   40.0  0.003333
36   36.6  0.003333
37   39.9  0.003333
38   35.4  0.003333
39   39.7  0.003333
40   36.1  0.003333

pulse     index     pulse
0    48.0  0.093333
1    60.0  0.083333
2     NaN  0.080000
3    40.0  0.060000
4    52.0  0.040000
5    88.0  0.040000
6   

In [7]:
# ИЩЕМ частоту пропущенных категориальных значений 
print('surgery', data['surgery'].value_counts(dropna = False, normalize = True).reset_index())
print()
print('age', data['age'].value_counts(dropna = False, normalize = True).reset_index())
print()
print('case_num', data['case_num'].value_counts(dropna = False, normalize = True).reset_index())
print()
print('temp_of_extremities', data['temp_of_extremities'].value_counts(dropna = False, normalize = True).reset_index())
print('outcome', data['outcome'].value_counts(dropna = False, normalize = True).reset_index())

surgery    index   surgery
0    1.0  0.600000
1    2.0  0.396667
2    NaN  0.003333

age    index   age
0      1  0.92
1      9  0.08

case_num        index  case_num
0     529796  0.006667
1     529424  0.006667
2    5279822  0.006667
3     527916  0.006667
4     530526  0.006667
..       ...       ...
279   530255  0.003333
280   530254  0.003333
281   533836  0.003333
282   530251  0.003333
283   535043  0.003333

[284 rows x 2 columns]

temp_of_extremities    index  temp_of_extremities
0    3.0             0.363333
1    1.0             0.260000
2    NaN             0.186667
3    2.0             0.100000
4    4.0             0.090000
outcome    index   outcome
0    1.0  0.593333
1    2.0  0.256667
2    3.0  0.146667
3    NaN  0.003333


##### 'surgery' -  категориальные значения
Указывает, была ли проведена операция:
1 = Да, была операция.
2 = Лечение без операции

#есть одно пропущенное значение 
##### 'age' - категориальные значения
Указывет на воратную категорию обследуемой лошади:
1 = взрослая лошадь
2 = молодая < 6 месяцев

у нас 2 варианта значения : "1" соответсвует значению Adult (взрослая лошадь), второе значение должно быть "2", но второе встречающееся значение - "9" (не "2"), которое по логике предполагается для молодой лошади (Young)
поскольку третьего варианта не дано, будет считать, что под "9" скрывается значение "Young"

###### 'case_num' -  категориальные значения
Поскольку некоторые лошади обрабатываются более 1 раза, номера могут совпадать (из 300 - 284 уникальных номера)
номер дела нужно привести к категориальному признаку, поскольку метрики по нему мы считать не будем

###### 'rec_temp' -  количественные значения (нормальная температура 37,8)
Указывается в градусах Цельсия. 
Повышенная температура может быть индикатором инфекции
Температура может быть пониженой, когда животное находится в позднем шоковом состоянии.
нормальная температура 37,8
Этот параметр обычно изменяется по мере того, как проблема прогрессирует. Например, может быть нормальной, а затем подняться из-за того, что поражение происходило в диапазоне, когда лошадь впала в шок

имеется 60 пропущенных значений (из 300)

###### 'pulse' -  количественные значения (30 -40 нормально для взрослых лошадей)
частота пульса в минуту
является показателем состояния сердца: 30 -40 нормально для взрослых
редко имеют более низкий уровень, чем обычно, хотя спортивные лошади может иметь 20-25
лошадь с болезненными повреждениями или страдающие от шока кровообращения может иметь повышенный пульс

имеется 24 пропущенных значений (из 300)"
                       
###### 'temp_of_extremities' - категориальные значения (холодные - шок, горячие - должны коррелировать с ректальной температурой)
субъективное указание на периферическое кровообращение
Возможные значения:
1 = Норма
2 = Тепло
3 = Холодно
4 = Слишком холодно
холодные или слишком холодные конечности указывают на возможный шок
горячие конечности должны коррелировать с повышенной ректальной температурой

имеется 56 пропущенных значений (из 300)

###### 'total_protein'  -  количественные значения (норма - 6-7,5)
нормальные значения лежат в диапазоне 6-7,5 (gms/dL).
чем выше значение, тем больше обезвоживание

имеется 33 пропущенных значений (из 300)

###### 'outcome'  - категориальные значения
что в конце концов случилось с лошадью?
Возможные значения:
1 = жива
2 = умерла
3 = была эвтаназирована

имеется только одно пропущенное значение

In [8]:
# меняем цифры словами, указывающими на определенные категории
data.loc[data['surgery'] == 1, 'surgery'] = 'Yes'
data.loc[data['surgery'] == 2, 'surgery'] = 'No'
data.loc[data['age'] == 1, 'age'] = 'Adult'
data.loc[data['age'] == 9, 'age'] = 'Young'
data.loc[data['temp_of_extremities'] == 1, 'temp_of_extremities'] = 'Normal'
data.loc[data['temp_of_extremities'] == 2, 'temp_of_extremities'] = 'Warm'
data.loc[data['temp_of_extremities'] == 3, 'temp_of_extremities'] = 'Cool'
data.loc[data['temp_of_extremities'] == 4, 'temp_of_extremities'] = 'Cold'
data.loc[data['outcome'] == 1, 'outcome'] = 'lived'
data.loc[data['outcome'] == 2, 'outcome'] = 'died'
data.loc[data['outcome'] == 3, 'outcome'] = 'was euthanized'

data.head()

surgery    age  case_num  rec_temp  pulse temp_of_extremities  \
0      No  Adult    530101      38.5   66.0                Cool   
1     Yes  Adult    534817      39.2   88.0                 NaN   
2      No  Adult    530334      38.3   40.0              Normal   
3     Yes  Young   5290409      39.1  164.0                Cold   
4      No  Adult    530255      37.3  104.0                 NaN   

   total_protein         outcome  
0            8.4            died  
1           85.0  was euthanized  
2            6.7           lived  
3            7.2            died  
4            7.4            died

In [9]:
# номер дела нужно привести к категориальному признаку, поскольку метрики по нему мы считать не будем
data['case_num'] = data.case_num.astype('object')

In [10]:
#Проводим рассчет базовых метрик
# все методы pandas по-умолчанию просто не берут в расчет пропуски
# В нашем случае вместо пропусков стоят "?"
# Чтобы произвести необходимые расчеты, нам нужно чтобы значения были числами или хотя бы пропусками
print('Базовые метрики (без учета пропусков):')
print()
print('Ректальная температура (rec_temp):')
print()
print('среднее значение:', data.rec_temp.mean())
print('медиана:', data.rec_temp.median())
print('мода:', data.rec_temp.mode()[0])
print('СКО:', data.rec_temp.std())
print('дисперсия:', data.rec_temp.var())
print()
print('Пульс (pulse):')
print()
print('среднее значение:', data.pulse.mean())
print('медиана:', data.pulse.median())
print('мода:', data.pulse.mode()[0])
print('СКО:', data.pulse.std())
print('дисперсия:', data.pulse.var())
print()
print('Общий белок (total_protein):')
print()
print('среднее значение:', data.total_protein.mean())
print('медиана:', data.total_protein.median())
print('мода:', data.total_protein.mode()[0])
print('СКО:', data.total_protein.std())
print('дисперсия:', data.total_protein.var())
print()
# мы можем вывести моду для категориальных признаков
print('Мода категориальных значений (без учета пропусков):')
print()
print('Мода surgery:', data.surgery.mode()[0])
print('Мода age:', data.age.mode()[0])
print('Мода temp_of_extremities:', data.temp_of_extremities.mode()[0])
print('Мода outcome:', data.outcome.mode()[0])

Базовые метрики (без учета пропусков):

Ректальная температура (rec_temp):

среднее значение: 38.16791666666669
медиана: 38.2
мода: 38.0
СКО: 0.7322886641121578
дисперсия: 0.5362466875871686

Пульс (pulse):

среднее значение: 71.91304347826087
медиана: 64.0
мода: 48.0
СКО: 28.630556660735003
дисперсия: 819.7087747035575

Общий белок (total_protein):

среднее значение: 24.456928838951317
медиана: 7.5
мода: 6.5
СКО: 27.475009470785064
дисперсия: 754.8761454197289

Мода категориальных значений (без учета пропусков):

Мода surgery: Yes
Мода age: Adult
Мода temp_of_extremities: Cool
Мода outcome: lived


In [11]:
# смотрим базовые метрики с помощью describe
data.describe()

rec_temp       pulse  total_protein
count  240.000000  276.000000     267.000000
mean    38.167917   71.913043      24.456929
std      0.732289   28.630557      27.475009
min     35.400000   30.000000       3.300000
25%     37.800000   48.000000       6.500000
50%     38.200000   64.000000       7.500000
75%     38.500000   88.000000      57.000000
max     40.800000  184.000000      89.000000

### Описание результатов количественных показателей

#### Ректальная температура:
Ректальная температура лошадей варируется от 35.4 С до 40.8 С. 
Учитывая, нормой является 37,8 и что среднее значение 38.1, делаем вывод о том, что выбросы, если они есть, скорее всего незначительны.
60 пропущенных значений (20 %)
Решение: 
Заполнить пропуски медианой
#### Пульс:
Пульс лошадей варируется от 30 до 184. 
Учитывая, что нормой является 30-40 , а среднее значение: 71.9 ( при медиане = 64.0 и моде = 48.0) то можно сделать вывод, что  что имеются небольшие выбросы в данных и что в основном наблюдаемые лошади имеют болезненные повреждения или страдают от шока кровообращения.
24 пропущенных значений (6 %)
Решение:
Заполнить пропуски медианой 
#### Общий белок:
Общий белок в анализах лошадей варируется от 3.3 до 89
Учитывая, что нормальные значения лежат в диапазоне 6-7,5 (gms/dL), а среднее значение = 24.45 (при медиане = 7.5 и моде = 6.5), то можно сделать выводы, что имеются серьезные выбросы в данных
33 пропущенных значений (11 %)
Решение: 
Заполнить пропуски медианой 

### Описание результатов категориальных показателей

#### Хирургическая операция (surgery):
1 пропущенное значение (< 1 %)
Решение: 
Заполнить пропуск модой
#### Возраст (age) :
Нет пропущенных значений (0 %)
#### Номер дела (case_num) :
Нет пропущенных значений (0 %)
#### Температура конечностей (temp_of_extremities):
56 пропущенных значений (18 %)
Решение: 
Заполнить пропуски модой
Обоснование :В комментариях к датасету сказано, что горячие конечности (Warm) должны коррелировать с повышенной ректальной температурой лошади, по факту в датасете присутствует большое количество значений противоречащих этому утверждению, во многих случаях, когда имеющиеся значения ректальной температуры, превышающие 39 C, соответствуют категории температуры конечностей, определенных как Normal или Cool, что является явным несоответсвием
#### Исход (outcome):
1 пропущенное значений (< 1 %)
Решение: 
Заполнить пропуск модой 

### Работаем с заменами пустых значений

In [12]:
data.rec_temp.fillna(data.rec_temp.median(), inplace=True)
data.pulse.fillna(data.pulse.median(), inplace=True)
data.total_protein.fillna(data.total_protein.median(), inplace=True)
data.describe()

rec_temp       pulse  total_protein
count  300.000000  300.000000     300.000000
mean    38.174333   71.280000      22.591667
std      0.654831   27.541545      26.453849
min     35.400000   30.000000       3.300000
25%     37.900000   48.000000       6.600000
50%     38.200000   64.000000       7.500000
75%     38.500000   88.000000      53.250000
max     40.800000  184.000000      89.000000

In [13]:
data.surgery.fillna(data.surgery.mode()[0], inplace=True)
data.outcome.fillna(data.outcome.mode()[0], inplace=True)
data.temp_of_extremities.fillna(data.temp_of_extremities.mode()[0], inplace=True)

### Работаем с выбросами количественных показателей
### Находим выбросы через межкваритльный размах

In [15]:
# считаем выбросы показателя rec_temp после замены пропусков значений (у нас было 60 пропущенных из 300):
# видим, что у нас 280 строк осталось после удаления выбросов:
q1 = data.rec_temp.quantile(0.25)
q3 = data.rec_temp.quantile(0.75)
iqr = q3 - q1
print('iqr (rec_temp) :', iqr)
lower_bound = q1 - (1.5 * iqr)
print('lower_bound :', lower_bound)
upper_bound = q3 + (1.5 * iqr)
print('upper_bound :', upper_bound)
data_2 = data[data.rec_temp.between(lower_bound, upper_bound, inclusive=True)].sort_values('rec_temp')
data_2

iqr (rec_temp) : 0.6000000000000014
lower_bound : 37.0
upper_bound : 39.400000000000006


surgery    age case_num  rec_temp  pulse temp_of_extremities  \
292     Yes  Adult   530034      37.0   66.0              Normal   
70       No  Adult   528214      37.0   60.0                Cool   
144     Yes  Adult   534519      37.1   84.0                Cool   
190     Yes  Adult   528742      37.1   40.0                Cool   
263     Yes  Adult   528743      37.1   75.0                Cool   
..      ...    ...      ...       ...    ...                 ...   
269     Yes  Adult   527883      39.3   64.0                Warm   
151     Yes  Adult   529399      39.3   64.0                Cold   
19       No  Adult   532110      39.4  110.0                Cold   
165     Yes  Adult   534925      39.4  120.0                Cool   
120     Yes  Adult   528682      39.4   54.0              Normal   

     total_protein         outcome  
292            6.9            died  
70             7.6  was euthanized  
144           81.0  was euthanized  
190            6.7           lived  
263            7.4            died  
..             ...             ...  
269            6.7           lived  
151            7.5            died  
19             8.7           lived  
165           64.0  was euthanized  
120            6.0           lived  

[280 rows x 8 columns]

### Мы нашли верхние выбросы в rec_tem - начиная от 39.4 С до максимального значения - 40.8 С
### Также мы нашли нижние выбросы от 35.4 С до минимального значения - 37.0 С
##### Чтобы оценить их влияние на базовые метрики проведем расчеты
Как видим далее самое чувствительное к выбросам - среднеарифмерическое (старое значение 38.174333 и новое значение 38.166071) поменялось незначительно. Решение - не удалять эти выбросы, поскольку они с большой вероятностью являются исключительными наблюдениями в данных, с меньшей уверенностью можно говорить о погрешности измерительных приборов

In [16]:
data.describe()

rec_temp       pulse  total_protein
count  300.000000  300.000000     300.000000
mean    38.174333   71.280000      22.591667
std      0.654831   27.541545      26.453849
min     35.400000   30.000000       3.300000
25%     37.900000   48.000000       6.600000
50%     38.200000   64.000000       7.500000
75%     38.500000   88.000000      53.250000
max     40.800000  184.000000      89.000000

In [17]:
# считаем выбросы показателя pulse после замены пропусков значений (до этого у нас было 24 пропущенных из 300):
# видим, что у нас 295 строк осталось после удаления выбросов:
q1 = data.pulse.quantile(0.25)
q3 = data.pulse.quantile(0.75)
iqr = q3 - q1
print('iqr (pulse) :', iqr)
lower_bound = q1 - (1.5 * iqr) 
print('lower_bound :', lower_bound)
upper_bound = q3 + (1.5 * iqr)
print('upper_bound  :', upper_bound)
data = data[data.pulse.between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
data

iqr (pulse) : 40.0
lower_bound : -12.0
upper_bound  : 148.0


surgery    age case_num  rec_temp  pulse temp_of_extremities  \
232     Yes  Adult  5262543      38.5   30.0                Cool   
130     Yes  Adult   528964      37.8   30.0                Cool   
203      No  Adult   529685      37.2   36.0              Normal   
276     Yes  Adult   534787      38.0   36.0                Cool   
242     Yes  Adult   530354      38.2   36.0              Normal   
..      ...    ...      ...       ...    ...                 ...   
148     Yes  Adult   529272      38.3  132.0                Cool   
135     Yes  Young  5287179      38.1  136.0                Cool   
44      Yes  Adult   535407      35.4  140.0                Cool   
103     Yes  Young  5292489      38.0  140.0              Normal   
39      Yes  Young  5277409      39.2  146.0                Cool   

     total_protein         outcome  
232            7.7           lived  
130            7.5            died  
203            5.7           lived  
276           75.0  was euthanized  
242            5.9            died  
..             ...             ...  
148            8.0           lived  
135            4.9            died  
44            69.0  was euthanized  
103            5.3           lived  
39             7.5            died  

[295 rows x 8 columns]

### Мы нашли верхние выбросы в pulse - начиная от 148 до максимального в нашем датафрейме - 184.00
Всего 5 значений
Решаем выбросы убрать, поскольку с большой вероятностью выбросы говорят об ошибке ввода данных (5 случаев на 300  - вполне возможно, сами ишибки тоже довольно грубые, учитывая, что норма варируется от 30 до 40), с меньшей уверенностью можно говорить о погрешности измерительных приборов

In [18]:
data.describe()

rec_temp       pulse  total_protein
count  295.000000  295.000000     295.000000
mean    38.162712   69.749153      22.863729
std      0.653795   25.053917      26.593188
min     35.400000   30.000000       4.000000
25%     37.900000   48.000000       6.600000
50%     38.200000   64.000000       7.500000
75%     38.400000   87.000000      54.000000
max     40.800000  146.000000      89.000000

In [19]:
# считаем выбросы показателя total_protein до замены пропусков значений (у нас 33 пропущенных из 300):
# видим, что у нас 267 строк осталось после удаления выбросов:
q1 = data.total_protein.quantile(0.25)
q3 = data.total_protein.quantile(0.75)
iqr = q3 - q1
print('iqr (total_protein) :', iqr)
lower_bound = q1 - (1.5 * iqr)
print('lower_bound (total_protein) :', lower_bound)
upper_bound = q3 + (1.5 * iqr)
print('upper_bound (total_protein) :', upper_bound)
data = data[data.total_protein.between(lower_bound, upper_bound, inclusive=True)].sort_values('total_protein')
data

iqr (total_protein) : 47.4
lower_bound (total_protein) : -64.5
upper_bound (total_protein) : 125.1


surgery    age case_num  rec_temp  pulse temp_of_extremities  \
210     Yes  Adult   530401      37.9   68.0                Cool   
142     Yes  Adult   530402      38.2  120.0                Cold   
84      Yes  Adult   529849      37.8   60.0              Normal   
247     Yes  Adult   530170      38.1   88.0                Cool   
191      No  Young  5287279      38.2  100.0                Warm   
..      ...    ...      ...       ...    ...                 ...   
144     Yes  Adult   534519      37.1   84.0                Cool   
63       No  Adult   534293      38.2  130.0                Cold   
1       Yes  Adult   534817      39.2   88.0                Cool   
90       No  Adult   534197      38.0   52.0                Cool   
208     Yes  Adult   535338      37.8   88.0                Cool   

     total_protein         outcome  
210            4.0            died  
142            4.5            died  
84             4.5           lived  
247            4.6            died  
191            4.7           lived  
..             ...             ...  
144           81.0  was euthanized  
63            82.0  was euthanized  
1             85.0  was euthanized  
90            86.0           lived  
208           89.0  was euthanized  

[295 rows x 8 columns]

### Мы не нашли выбросов в total_protein - у нас не встречается значений, превышающих 125.1 gms/dL

In [20]:
data.describe()

rec_temp       pulse  total_protein
count  295.000000  295.000000     295.000000
mean    38.162712   69.749153      22.863729
std      0.653795   25.053917      26.593188
min     35.400000   30.000000       4.000000
25%     37.900000   48.000000       6.600000
50%     38.200000   64.000000       7.500000
75%     38.400000   87.000000      54.000000
max     40.800000  146.000000      89.000000

### Нормальный датафрейм без пустых значений

In [21]:
# Всего в датафрейме 295 строчек
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 295 entries, 210 to 208
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   surgery              295 non-null    object 
 1   age                  295 non-null    object 
 2   case_num             295 non-null    object 
 3   rec_temp             295 non-null    float64
 4   pulse                295 non-null    float64
 5   temp_of_extremities  295 non-null    object 
 6   total_protein        295 non-null    float64
 7   outcome              295 non-null    object 
dtypes: float64(3), object(5)
memory usage: 20.7+ KB
